lib/cli/src/Cardano/CLI.hs

```diff
@@ -500,6 +503,48 @@ decodeKey = \case
       where
         dpErr = "Internal error: Unable to extract bytes from bech32 data part"
 
+        bech32Err  = ("Bech32 error: " <>) . \case
+            StringToDecodeTooLong -> "string is too long"
+            StringToDecodeTooShort -> "string is too short"
+            StringToDecodeHasMixedCase -> "mixed case"
+            StringToDecodeMissingSeparatorChar -> "missing separator char"
+            StringToDecodeContainsInvalidChars ixs ->
+                let ixs' = map unCharPos ixs
+                in mconcat
+                    [ "Invalid character(s) in string:\n"
+                    , markCharsRedAtIndices ixs' (T.unpack t)
+                    , "\n"
+                    , arrowsAtIndices $ map unCharPos ixs
+                    ]
+        unCharPos (CharPosition x) = x
+
+markCharsRedAtIndices :: Integral i => [i] -> String -> String
+markCharsRedAtIndices ixs txt = go 0 ixs txt ++ def
+  where
+    go _c [] [] = mempty
+    go c (i:is) (s:ss)
+        | c == i    = red ++ s:def ++ go (c + 1) is ss
+        | otherwise = s : go (c + 1) (i:is) ss
+    go _ [] ss = ss
+    go _ _ [] = [] -- NOTE: Really an error case.
+
+    red = setSGRCode [SetColor Foreground Vivid Red]
+    def = setSGRCode [Reset]
+
+-- E.g.
+-- >>> arrowsAtIndices [0,2]
+-- >>> "^ ^"
+arrowsAtIndices :: Integral i => [i] -> String
+arrowsAtIndices ixs = go 0 ixs ++ def
+  where
+    go _c [] = mempty
+    go c (x:xs)
+        | c == x    = red ++ "^" ++ go (c + 1) xs
+        | otherwise = ' ' : go (c + 1) (x:xs)
+
+    red = setSGRCode [SetColor Foreground Vivid Red]
+    def = setSGRCode [Reset]
```

Comment:
See: https://input-output-hk.github.io/cardano-wallet/haddock/edge/cardano-wallet-cli-2020.4.28/src/Cardano.CLI.html#withSGR which also properly handles non-terminal environments :)

Reply:
I'm not in IO here though 🤔

Reply:
Well, yeah, the function needs IO to determines whether the code runs in a terminal or not.